compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -525,53 +529,189 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
+                    };
+                }
+
+                let niche_end =
+                    tag_size.truncate(u128::from(relative_max).wrapping_add(niche_start));
+
+                // Next, the layout algorithm prefers to put the niches at one end,
+                // so look for cases where we don't need to calculate a relative_tag
+                // at all and can just look at the original tag value directly.
+                // This also lets us move any possibly-wrapping addition to the end
+                // where it's easiest to get rid of in the normal uses: it's easy
+                // to optimize `COMPLICATED + 2 == 7` to `COMPLICATED == (7 - 2)`.
+                {
+                    // Work in whichever size is wider, because it's possible for
+                    // the untagged variant to be further away from the niches than
+                    // is possible to represent in the smaller type.
+                    let (wide_size, wide_ibty) = if cast_to_layout.size > tag_size {
+                        (cast_to_layout.size, cast_to)
+                    } else {
+                        (tag_size, tag_llty)
+                    };
+
+                    struct NoWrapData<V> {
+                        wide_tag: V,
+                        is_niche: V,
+                        needs_assume: bool,
+                        wide_niche_to_variant: u128,
+                        wide_niche_untagged: u128,
                     }
 
-                    (is_niche, cast_tag, niche_variants.start().as_u32() as u128)
-                };
+                    let first_variant = u128::from(niche_variants.start().as_u32());
+                    let untagged_variant = u128::from(untagged_variant.as_u32());
+
+                    let opt_data = if tag_range.no_unsigned_wraparound(tag_size) == Ok(true) {
+                        let wide_tag = bx.zext(tag, wide_ibty);
+                        let extend = |x| x;
+                        let wide_niche_start = extend(niche_start);
+                        let wide_niche_end = extend(niche_end);
+                        debug_assert!(wide_niche_start <= wide_niche_end);
+                        let wide_first_variant = extend(first_variant);
+                        let wide_untagged_variant = extend(untagged_variant);
+                        let wide_niche_to_variant =
+                            wide_first_variant.wrapping_sub(wide_niche_start);
+                        let wide_niche_untagged = wide_size
+                            .truncate(wide_untagged_variant.wrapping_sub(wide_niche_to_variant));
+                        let (is_niche, needs_assume) = if tag_range.start == niche_start {
+                            let end = bx.cx().const_uint_big(tag_llty, niche_end);
+                            (
+                                bx.icmp(IntPredicate::IntULE, tag, end),
+                                wide_niche_untagged <= wide_niche_end,
+                            )
```

Comment:
I'm not sure how to review this, the amount of similarly named variables just overflows my cache :(